File: domains/DomainAgent.js
import axios from 'axios';

/**
 * DomainAgent integrates with Namecheap's API to manage DNS records.
 * This class handles creating and verifying _atproto TXT records for
 * automating handle verification via the AT Protocol.
 */
export default class DomainAgent {
  /**
   * @param {object} options
   * @param {string} options.apiUser - Namecheap ApiUser
   * @param {string} options.apiKey - Namecheap ApiKey
   * @param {string} options.clientIp - Authorized client IP for requests
   * @param {string} [options.baseUrl] - Base URL for Namecheap API
   */
  constructor({ apiUser, apiKey, clientIp, baseUrl = 'https://api.namecheap.com/xml.response' }) {
    this.apiUser = apiUser;
    this.apiKey = apiKey;
    this.clientIp = clientIp;
    this.baseUrl = baseUrl;
  }

  /**
   * Internal helper to send requests to Namecheap's API.
   */
  async _request(params) {
    const searchParams = new URLSearchParams({
      ApiUser: this.apiUser,
      ApiKey: this.apiKey,
      UserName: this.apiUser,
      ClientIp: this.clientIp,
      ...params,
    });
    const url = `${this.baseUrl}?${searchParams.toString()}`;
    const res = await axios.get(url, { responseType: 'text' });
    return res.data;
  }

  /**
   * Fetch existing DNS host records for a domain.
   */
  async getHosts(domain) {
    const [SLD, TLD] = domain.split('.');
    const xml = await this._request({ Command: 'namecheap.domains.dns.getHosts', SLD, TLD });
    return xml;
  }

  /**
   * Set all DNS host records for a domain.
   * `records` should be an array of objects like:
   * { HostName, RecordType, Address, TTL }
   */
  async setHosts(domain, records) {
    const [SLD, TLD] = domain.split('.');
    const base = {
      Command: 'namecheap.domains.dns.setHosts',
      SLD,
      TLD,
    };
    records.forEach((rec, i) => {
      base[`HostName${i + 1}`] = rec.HostName;
      base[`RecordType${i + 1}`] = rec.RecordType;
      base[`Address${i + 1}`] = rec.Address;
      base[`TTL${i + 1}`] = rec.TTL || '60';
    });
    const xml = await this._request(base);
    return xml;
  }

  /**
   * Convenience to add a TXT record.
   */
  async addTxtRecord(domain, host, value, ttl = 60) {
    const current = await this.getHosts(domain);
    // Minimal XML parsing to extract hosts is omitted for brevity.
    // In real code you would parse XML to objects. Here we expect caller
    // to provide full records list.
    throw new Error('addTxtRecord requires existing host parsing not implemented');
  }

  /**
   * Configure the _atproto TXT record for the provided DID.
   */
  async configureHandle(domain, did) {
    const record = {
      HostName: '_atproto',
      RecordType: 'TXT',
      Address: `did=${did}`,
      TTL: '60',
    };
    // For simplicity we replace all existing records with this single record.
    await this.setHosts(domain, [record]);
  }
}

import dns from 'dns/promises';

/**
 * Poll DNS for _atproto TXT record to verify propagation.
 * @param {string} domain
 * @param {string} did
 * @param {number} [retries=10]
 * @param {number} [delay=30000] milliseconds between polls
 */
export async function verifyRecord(domain, did, retries = 10, delay = 30000) {
  const expected = `did=${did}`;
  for (let i = 0; i < retries; i++) {
    try {
      const records = await dns.resolveTxt(`_atproto.${domain}`);
      const flat = records.flat().join('');
      if (flat.includes(expected)) return true;
    } catch (err) {
      // ignore DNS errors
    }
    await new Promise(r => setTimeout(r, delay));
  }
  return false;
}

File: forge_domains/domain_agent.py
import os
import requests
import time
import dns.resolver
from urllib.parse import urlencode

class DomainAgent:
    NAMECHEAP_API_URL = "https://api.namecheap.com/xml.response"

    def __init__(self, api_user=None, api_key=None, client_ip=None, username=None):
        self.api_user = api_user or os.getenv("NAMECHEAP_API_USER")
        self.api_key = api_key or os.getenv("NAMECHEAP_API_KEY")
        self.client_ip = client_ip or os.getenv("NAMECHEAP_CLIENT_IP")
        self.username = username or os.getenv("NAMECHEAP_USERNAME", self.api_user)
        if not all([self.api_user, self.api_key, self.client_ip]):
            raise ValueError("Missing Namecheap API credentials")

    def _request(self, command, params):
        query = {
            "ApiUser": self.api_user,
            "ApiKey": self.api_key,
            "UserName": self.username,
            "ClientIp": self.client_ip,
            "Command": command,
            **params,
        }
        response = requests.get(self.NAMECHEAP_API_URL, params=query)
        response.raise_for_status()
        return response.text

    def set_txt_record(self, domain, host, value, ttl=60):
        return self._request(
            "namecheap.domains.dns.setHosts",
            {
                "SLD": domain.split(".")[0],
                "TLD": domain.split(".")[1],
                "HostName1": host,
                "RecordType1": "TXT",
                "Address1": value,
                "TTL1": ttl,
            },
        )

    def verify_dns(self, fqdn, expected_value, timeout=300):
        end_time = time.time() + timeout
        while time.time() < end_time:
            try:
                answers = dns.resolver.resolve(fqdn, "TXT")
                for rdata in answers:
                    if expected_value in str(rdata.strings[0], "utf-8"):
                        return True
            except Exception:
                pass
            time.sleep(30)
        return False

if __name__ == "__main__":
    import sys

    if len(sys.argv) != 3:
        print("Usage: domain_agent.py <domain> <did>")
        sys.exit(1)

    domain, did = sys.argv[1], sys.argv[2]
    agent = DomainAgent()
    agent.set_txt_record(domain, "_atproto", f"did={did}")
    fqdn = f"_atproto.{domain}"
    if agent.verify_dns(fqdn, f"did={did}"):
        print(f"{domain} verified for {did}")
    else:
        print("Verification timed out")
        sys.exit(1)
